### ISSUE TITLE
Libraries contain TEXTRELs

### ISSUE THREAD

**Issue by andreas-abel** [lkl/linux#220]
The LKL libraries contain TEXTRELs (i.e., relocations in the text segment). This can be seen using "readelf -d liblkl.so | grep TEXTREL". 

Having TEXTRELs can lead to a number of issues:

1. If a shared library contains a TEXTREL, the corresponding memory pages cannot be shared.
2. Android >=6 and SELinux do not support binaries with TEXTRELs (http://android-developers.blogspot.com/2016/06/android-changes-for-ndk-developers.html).
3. If a program has a TEXTREL, uses an ifunc, and is compiled with early binding, this can lead to a segmentation fault when processing the relocation for the ifunc during dynamic linking because the text segment is made temporarily non-executable to process the TEXTREL (line 248 in dl_reloc.c).

The TEXTREL is caused by the assembler code that is generated by kallsyms.c. The problem here is that this data is put into the .rodata section, which ends up in the same segment as the .text section.

One straightforward fix is to move this data to the (writable) .data section instead (by modifying line 355 in kallsyms.c correspondingly; this was actually the original behavior until kernel 2.6.19 (http://lkml.iu.edu/hypermail/linux/kernel/0612.0/0988.html)).

**Comment by andreas-abel** [comment]
Here is an example for issue #3 that leads to a segmentation fault:

bug.c:
--------------------------------------------------------------------------------
#include <lkl.h>

#include <stdio.h>
#include <sys/auxv.h>

int foo_v4 (void) { return 4; }

void * f_choice (void) {
  return foo_v4;
}

int foo (void) __attribute__ ((ifunc ("f_choice")));

int main() {
  printf("vfpv%d\n", foo());
  volatile int b = 0;
  if (b) {
    // never executed
    lkl_syscall(0, 0);
  }
  return 0;
}
--------------------------------------------------------------------------------

Compile with (gcc version 5.3.0):
gcc bug.c liblkl.a -Iinclude -o bug -lpthread -lrt  -Wl,-z,noexecstack -fPIC  -pie

**Comment by andreas-abel** [comment]
Here is an example for issue #3 that leads to a segmentation fault:

bug.c:
-------------------------------------------------------
#include <lkl.h>
#include <stdio.h>
#include <sys/auxv.h>

int foo_v4 (void) { return 4; }

void * f_choice (void) {
  return foo_v4;
}

int foo (void) __attribute__ ((ifunc ("f_choice")));

int main() {
  printf("vfpv%d\n", foo());
  volatile int b = 0;
  if (b) {
    // never executed
    lkl_syscall(0, 0);
  }
  return 0;
}
-------------------------------------------------------

Compile with (gcc version 5.3.0):
gcc bug.c liblkl.a -Iinclude -o bug -lpthread -lrt -Wl,-z,noexecstack -fPIC -pie

**Comment by andreas-abel** [comment]
Here is an example for issue #3 that leads to a segmentation fault:

bug.c:
```
#include <lkl.h>
#include <stdio.h>
#include <sys/auxv.h>

int foo_v4 (void) { return 4; }

void * f_choice (void) {
  return foo_v4;
}

int foo (void) __attribute__ ((ifunc ("f_choice")));

int main() {
  printf("vfpv%d\n", foo());
  volatile int b = 0;
  if (b) {
    // never executed
    lkl_syscall(0, 0);
  }
  return 0;
}
```

Compile with (gcc version 5.3.0):
gcc bug.c liblkl.a -Iinclude -o bug -lpthread -lrt -Wl,-z,noexecstack -fPIC -pie

**Comment by liuyuan10** [comment]
@opurdila We encountered seg fault when using the liblkl.a in some applications. Andreas helped us find out the root cause and a workaround. The fix is vital to us and has been used in our repository. I think it may be useful for upstream. But the downside is the fix needs to touch kernel kallsyms.c. What do you think?

**Comment by opurdila** [comment]
@andreas-abel thanks for investigating and finding a fix to the issue !

I think we might be able to get around modifying the kallsyms with some objcopy magic, but since this is a critical fix for you and it is very simple, lets merge the fix you have for now.

**Comment by liuyuan10** [comment]
Just to keep a record. Another options without touching kallsyms.c is to disable both CONFIG_KALLSYMS and CONFIG_KALLSYMS_ALL at the cost of loosing some debuggability. @andreas-abel could you send out a PR of the fix?

**Comment by opurdila** [comment]
Fixed by #221
